File: java/78_Subsets.java
//sol binary
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        int len = nums.length;

        for (int i = (int)Math.pow(2, len); i < (int)Math.pow(2, len+1); i++) {
            String i2b = Integer.toBinaryString(i).substring(1);
            
            List<Integer> cur = new ArrayList<>();

            for (int j = 0; j < len; j++) {
                if (i2b.charAt(j) == '1') {
                    cur.add(nums[j]);
                }
            }
            res.add(cur);
        }
        return res;
    }
}
//sol dfs
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        res.add(new ArrayList<Integer>());

        for (int num : nums) {
            List<List<Integer>> curSubsets = new ArrayList<>();
            for (List<Integer> subset : res) {
                List<Integer> curSubset =  new ArrayList<>(subset);
                curSubset.add(num);
                curSubsets.add(curSubset);
            }
            res.addAll(curSubsets);
        }

        return res;
    }
}

File: c++/78_Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        int n = nums.size();
        for (int i = 0; i < (1 << n); i++) {
            vector<int> path;
            for (int j = 0; j < n; j++) {
                if ((i >> j) & 1) path.push_back(nums[j]);
            }
            res.push_back(path);
        }
        return res;
    }
};
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    
    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(nums, 0);
        return res;
    }
    
    void dfs(vector<int>& nums, int u) {
        if (u == nums.size()) {
            res.push_back(path);
            return;
        }
        path.push_back(nums[u]);
        dfs(nums, u+1);
        path.pop_back();
        dfs(nums, u+1);
    }
};
